src/core/lib/gpr/fork.cc

```diff
@@ -69,10 +84,104 @@ void grpc_fork_support_init() {
   if (override_fork_support_enabled != -1) {
     fork_support_enabled = override_fork_support_enabled;
   }
+  if (fork_support_enabled) {
+    g_fork_state = (fork_state*)gpr_malloc(sizeof(fork_state));
+    gpr_mu_init(&g_fork_state->mu);
+    gpr_cv_init(&g_fork_state->thd_cv);
+    gpr_cv_init(&g_fork_state->exec_ctx_cv);
+    gpr_atm_no_barrier_store(&g_fork_state->exec_ctx_count, 2);
+    g_fork_state->thd_count = 0;
+    g_fork_state->awaiting_threads = false;
+  }
+}
+
+void grpc_fork_support_destroy() {
+  if (g_fork_state) {
+    gpr_mu_destroy(&g_fork_state->mu);
+    gpr_cv_destroy(&g_fork_state->thd_cv);
+    gpr_cv_destroy(&g_fork_state->exec_ctx_cv);
+    gpr_free(g_fork_state);
+    g_fork_state = nullptr;
+  }
+}
+
+void grpc_fork_inc_thd_count() {
+  if (g_fork_state) {
+    gpr_mu_lock(&g_fork_state->mu);
+    g_fork_state->thd_count++;
+    gpr_mu_unlock(&g_fork_state->mu);
+  }
+}
+
+void grpc_fork_dec_thd_count() {
+  if (g_fork_state) {
+    gpr_mu_lock(&g_fork_state->mu);
+    g_fork_state->thd_count--;
+    if (g_fork_state->awaiting_threads && g_fork_state->thd_count == 0) {
+      gpr_cv_signal(&g_fork_state->thd_cv);
+    }
+    gpr_mu_unlock(&g_fork_state->mu);
+  }
+}
+
+void grpc_fork_await_thds() {
+  if (g_fork_state) {
+    gpr_mu_lock(&g_fork_state->mu);
+    g_fork_state->awaiting_threads = true;
+    gpr_cv_wait(&g_fork_state->thd_cv, &g_fork_state->mu,
```

Comment:
We shouldn't ever have a `cv_wait` without a controlling boolean. It should always be of the form

```cpp

while (!done_) {
  gpr_cv_wait(&done_cv_, &mu_, gpr_inf_future....);
}
```

and on the other side

```cpp
gpr_mu_lock(&mu_);
done_ = true;
gpr_cv_signal_(&done_cv_);
gpr_mu_unlock(&mu_);
```

The reason is twofold. First, `cv_wait` can have spurious wakeups because of interrupts, etc. Second, if you don't check if the condition is already true before waiting, you might have already missed the signal.

Reply:
Good catch.  Fixed